### ISSUE TITLE
Feature ask: support in memory filtering after ODataQueryOptions.ApplyTo()

### ISSUE THREAD

**Issue by xubinzheng** [OData/AspNetCoreOData#766]
I want to follow up on this very old ticket: https://github.com/OData/WebApi/issues/521

This issue was raised in 2015, and the same limitation seems still exist in the Odata.ApplyTo(). I am wondering is there a better solution than `You can do this to check for a $select wrapper, get a Dictionary of key,values and then map it your model.` at this moment?
<img width="732" alt="image" src="https://user-images.githubusercontent.com/9353989/204999709-4f438aa4-3539-44fe-8abf-cb99da7ecd57.png">

Similar to the original question, I am trying to perform in memory filtering of a query, after it has been retrieved from the downstream API, and before returning it to the client.

Using ODataQueryOptions.ApplyTo(...) I can let OData do all it's query magic on the in-memory data. However if $expand or $select has been supplied, the returned IQueryable contains not the entities themselves, but SelectExpandWrapper<TElement> objects. Which it is impossible to extract the original entity from.

Ideally I would like to be able to do something like this in a controller action

```
public IQueryable<Entity> GetFeed(ODataQueryOptions<Entity> opts)
{
    IQueryable<Entity> result = opts.ApplyTo(<IEnumerable>.AsQueryable<Entity>())

    foreach (var item in result)
    {
        // perform extra filtering here
    }
    return result;
}

```

### Expected result

Any suggestions will be very helpful.

**Comment by julealgon** [comment]
Can't you filter the data before applying the queryoptions?

Can you elaborate a bit more on what is it exactly you are trying to do, and why you think filtering after applying the OData query options would be the way to go?

**Comment by xubinzheng** [comment]
@julealgon  yep. Firstly the `queryoptions` are applied to the downstream APIs directly, like calling MSGraph with `$select=Id`. So the data doesn't exist in the memory for filtering before applying the queryoptions.

More specifically, in our case, this is for an evaluation purpose. 

We have a legacy code path that calls some downstream APIs. And we have a new code path that calls MSGraph. Today when a request came in, we run both flows in parallel, and only return the response from the legacy flow. But before we return the response, we have the logic that compares results between the legacy and the new flow to make sure they are equivalent. That helps us smoothly switch to the new code path.

We need to cover the query results back to C# objects to apply the comparison logic.

**Comment by julealgon** [comment]
> @julealgon yep. Firstly the `queryoptions` are applied to the downstream APIs directly, like calling MSGraph with `$select=Id`. So the data doesn't exist in the memory for filtering before applying the queryoptions.

MSGraph is an OData API as well right? So you could just filter the data by appending extra filters on `$filter` before sending the request to them.

I'm not exactly sure how you are making that request, but I assume using a `DataServiceContext` would make the most sense: you could then add extra `Where` clauses to the context before applying the `ODataQueryOptions` onto it.

> More specifically, in our case, this is for an evaluation purpose.
> 
> We have a legacy code path that calls some downstream APIs. And we have a new code path that calls MSGraph. Today when a request came in, we run both flows in parallel, and only return the response from the legacy flow. But before we return the response, we have the logic that compares results between the legacy and the new flow to make sure they are equivalent. That helps us smoothly switch to the new code path.
> 
> We need to cover the query results back to C# objects to apply the comparison logic.

Would it not be better if you could filter the data in the legacy API upfront as well? Or is that just not an option for some reason?

You wouldn't need to apply any filters after any of the queries run in that case: just make the 2 calls with filtering already in place, and compare the final results directly.
